subsys/usb/class/cdc_acm.c

```diff
@@ -1118,4 +1118,19 @@ static const struct uart_driver_api cdc_acm_driver_api = {
 		CONFIG_KERNEL_INIT_PRIORITY_DEVICE,			\
 		&cdc_acm_driver_api);
 
+
+#define DT_DRV_COMPAT zephyr_cdc_acm_uart
+
+#define CDC_ACM_DT_DEVICE_DEFINE(idx)					\
+	CDC_ACM_CFG_AND_DATA_DEFINE(idx)				\
+									\
+	DEVICE_DT_INST_DEFINE(idx, cdc_acm_init, NULL,			\
+		&cdc_acm_dev_data_##idx, &cdc_acm_config_##idx,		\
+		POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE,	\
+		&cdc_acm_driver_api);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+DT_INST_FOREACH_STATUS_OKAY(CDC_ACM_DT_DEVICE_DEFINE);
+#else
 UTIL_LISTIFY(CONFIG_USB_CDC_ACM_DEVICE_COUNT, CDC_ACM_DEVICE_DEFINE, _)
+#endif
```

Comment:
Any reason not to allow both ways of defining these devices?

Reply:
I prefer only one option, as it is with other UART drivers for hardware controllers, and I would like to deprecate Kconfig/UTIL_LISTIFY option. I also see no advantage but a lot of conflict potential in the configuration for the user.

Reply:
I agree. It would be quite confusing if the user can specify both `CONFIG_USB_CDC_ACM_DEVICE_COUNT=2` and then add a 3rd USB CDC port via Devicetree. I think that keeping the old Kconfig config for backwards compatibility but switching to Devicetree entirely whenever it is used to define these ports makes more sense.

Reply:
In fact, ideally this PR would be extended to convert the in-tree samples that use Kconfig to DT, and deprecate the Kconfig options themselves. We shouldn't have two ways of achieving exactly the same objective, and DT is more flexible in that it allows the user to change settings per-instance, as well as being able to connect the `chosen` node for console, shell, logging, etc directly to a real DT node instead of falling back to Kconfig.

Reply:
I agree. Let's deprecate the Kconfig way of doing things.

@jfischer-no is USB CDC ACM the only reason we have CONFIG_UART_CONSOLE_ON_DEV_NAME?

If so, that should be deprecated too. That would help avoid the weird issues in the last patch, since we could:

- just get the console device out of devicetree directly with `DEVICE_DT_GET(DT_CHOSEN(zephyr_console))` to poll DTR
- verify the console is USB by using `BUILD_ASSERT(DT_NODE_HAS_COMPAT(DT_CHOSEN(zephyr_console)), zephyr_cdc_acm_uart), ...);`

While we are doing this, I wonder if we can deprecate the following Kconfig options in the same way if CDC ACM is the only thing keeping them around:

- CONFIG_BT_MONITOR_ON_DEV_NAME
- CONFIG_BT_UART_ON_DEV_NAME
- CONFIG_UART_SHELL_ON_DEV_NAME
- CONFIG_UART_PIPE_ON_DEV_NAME

Reply:
> @jfischer-no is USB CDC ACM the only reason we have CONFIG_UART_CONSOLE_ON_DEV_NAME?

There is also: `boards/nios2/altera_max10/doc/index.rst:   CONFIG_UART_CONSOLE_ON_DEV_NAME="jtag_uart0"`, `drivers/serial/uart_altera_jtag_hal.c`

Reply:
[subsys/bluetooth/host/monitor.c](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/monitor.c#L237) needs to be changed as well.